src/Endorser.sol

```diff
@@ -103,25 +127,40 @@ contract Endorser is
     function _validateAndNullifyProof(
         address to,
         bytes32 digest,
-        bytes32 root,
-        bytes32[] memory proof
+        Proof calldata proof,
+        address authorizer,
+        bytes calldata authorizerSignature
     ) internal {
-        bytes memory data = abi.encodeCall(Witness.witnessedAt, (root));
-        bytes32 leaf = _leaf(to, digest);
-        (, bytes32 timestamp) = _callReturnBytes32(address(witness()), data);
+        EndorserStorage storage $ = _getEndorserStorage();
+        bytes32 leaf = getProvenanceHash(abi.encode(to, digest));
+
+        // Sanity check
+        if (leaf != proof.leaf) revert MismatchedLeaf();
+
+        // Already nullified
+        if ($._nullifier.get(uint256(leaf))) revert AlreadyClaimed();
+
+        // Verify authorizer signature
         if (
-            !MerkleProof.verify(proof, root, leaf) || // invalid proof
-            _getEndorserStorage()._nullifier.get(uint256(leaf)) || // not nullified
-            timestamp == 0 // witnessed
-        ) {
-            revert InvalidProof();
-        }
-        _getEndorserStorage()._nullifier.set(uint256(leaf));
-    }
+            !SignatureChecker.isValidSignatureNow(
+                authorizer,
+                MessageHashUtils.toEthSignedMessageHash(leaf),
```

Comment:
I should change this message encoding to an EIP-712 message so that authorizations are chain-specific